_posts/2020-09-16-hibernate-search.md

```diff
@@ -0,0 +1,440 @@
+  fullTextQuery.setSort(qb.sort().byScore().createSort());
+  return (List<Post>) fullTextQuery.getResultList();
+}
+```
+
+As you can see in the snippet above for `likeCount` we are using range query. Using only `above()` is equivalent to `>=` 
+operator. If you want to exclude the limits, just call `excludeLimit()` after `above()`.
+  
+For the other two fields, we have again used a keyword query. 
+
+Now, it's time to combine all queries. To do so, we will make use of `QueryBuilder` object's `bool()` function which 
+provides us with verbs such as `should()`, `must()` and `not()`. I have used `must()` for `likeCount` query and `should()`
+for the rest as they are optional. Optional queries wrapped in `should()` contribute to the relevance score. For instance document with tag matching, 'Literature' will score higher.
+
+A couple of paragraphs ago I had promised you that I will show you a way to avoid queries to the database. So, here it is, and 
+it's called **Projection**.
+```java
+public List<User> getUserByFirstWithProjection(String first, 
+                                               int max, 
+                                               int page){
+
+  FullTextEntityManager fullTextEntityManager = 
+                        Search.getFullTextEntityManager(entityManager);
+  QueryBuilder qb = fullTextEntityManager
+            .getSearchFactory()
+            .buildQueryBuilder()
+            .forEntity(User.class)
+            .get();
+  
+  Query similarToUser = qb.keyword().fuzzy()
+            .withEditDistanceUpTo(2)
+            .onField("first")
+            .matching(first)
+            .createQuery();
+  Query finalQuery = qb.bool()
+            .must(similarToUser)
+            .createQuery();
+  FullTextQuery fullTextQuery = 
+            fullTextEntityManager.createFullTextQuery(finalQuery, User.class);
+  fullTextQuery.setProjection(FullTextQuery.ID, "first","last","middle","age");
+  fullTextQuery.setSort(qb.sort()
+                            .byField("age")
+                            .desc()
+                            .andByScore()
+                            .createSort());
+  fullTextQuery.setMaxResults(max);
+  fullTextQuery.setFirstResult(page);
+  return getUserList(fullTextQuery.getResultList());
+}
+
+private List<User> getUserList(List<Object[]> resultList) {
+  List<User> users = new ArrayList<>();
+  for (Object[] objects : resultList) {
+      User user = new User();
+      user.setId((String) objects[0]);
+      user.setFirst((String) objects[1]);
+      user.setLast((String) objects[2]);
+      user.setMiddle((String) objects[3]);
+      user.setAge((Integer) objects[4]);
+      users.add(user);
+  }
+  return users;
+}
+```
+
+Before moving on to the main event let's zoom in on a couple of interesting things here.
+
+* Up until now we were using keyword queries to perform exact match searches only, but here as you can see I have combined
```

Comment:
This is too much text for a list. A list means you can't separate the text into multiple paragraphs, so it becomes a wall of text. Please make this normal text and split it into meaningful paragraphs to reduce the wall of text.

Reply:
Actually, I think that "Fuzzy Search" and "Pagination" should each have their own subsection. Please think about creating a small code snippet for each, that focuses on each feature alone and creating a separate section for each. 

A section can very well be a short code snippet followed by a single paragraph. It doesn't need to be long